pkg/watch/manager_integration_test.go

```diff
@@ -0,0 +1,542 @@
+	return c.Watch(
+		&source.Channel{
+			Source:         events,
+			DestBufferSize: 1024,
+		},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: util.EventPacker{}},
+	)
+}
+
+// Verify that an unknown resource will return an error when adding a watch.
+func TestRegistrar_AddUnknown(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	mgr, wm := setupManager(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	grp, ctx := errgroup.WithContext(ctx)
+
+	grp.Go(func() error {
+		return mgr.Start(ctx.Done())
+	})
+
+	events := make(chan event.GenericEvent)
+	r, err := wm.NewRegistrar("test", events)
+	g.Expect(err).NotTo(gomega.HaveOccurred(), "creating registrar")
+
+	err = r.AddWatch(schema.GroupVersionKind{
+		Group:   "i",
+		Version: "donot",
+		Kind:    "exist",
+	})
+	g.Expect(err).To(gomega.HaveOccurred(), "AddWatch should have failed due to unknown GVK")
+
+	cancel()
+	_ = grp.Wait()
+}
+
+// Verifies that controller-runtime interleaves reconcile errors in backoff and
+// other events within the same work queue.
+func Test_ReconcileErrorDoesNotBlockController(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	mgr, _ := setupManager(t)
+	ctrl.SetLogger(logf.NullLogger{})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	grp, ctx := errgroup.WithContext(ctx)
+
+	grp.Go(func() error {
+		return mgr.Start(ctx.Done())
+	})
+
+	// Events will be used to receive events from dynamic watches registered
+	// via the registrar below.
+	errObj := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "error",
+		},
+	}
+	events := make(chan event.GenericEvent, 1024)
+	events <- event.GenericEvent{
+		Meta:   errObj,
+		Object: errObj,
+	}
+
+	requests := make(chan reconcile.Request)
+	rec := func(request reconcile.Request) (reconcile.Result, error) {
+		select {
+		case requests <- request:
+		case <-ctx.Done():
+		}
+
+		if request.Name == "error" {
+			return reconcile.Result{}, errors.New("synthetic error")
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Create a new controller
+	c, err := controller.New("test-controller", mgr, controller.Options{Reconciler: reconcile.Func(rec)})
+	if err != nil {
+		t.Fatalf("creating controller: %v", err)
+	}
+	err = c.Watch(
+		&source.Channel{
+			Source:         events,
+			DestBufferSize: 1024,
+		},
+		&handler.EnqueueRequestForObject{},
+	)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	// Wait for the error resource to reconcile
+	// before setting up another watch.
+	<-requests
+
+	// Setup another watch. Show that both the error resource (in a backoff-requeue loop)
+	// and other resources can reconcile in an interleaving fashion.
+	err = c.Watch(
+		&source.Kind{Type: &corev1.Namespace{}},
+		&handler.EnqueueRequestForObject{},
+	)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	expectNames := map[string]bool{"error": true, "default": true, "kube-system": true}
+loop:
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
what's the timeout for the manager to exit? Is there a timeout? Otherwise we risk an infinite loop.

Reply:
5 seconds, set here:

```
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Beyond that go tests have a timeout of 10 minutes by default:
```
        -timeout d
            If a test binary runs longer than duration d, panic.
            If d is 0, the timeout is disabled.
            The default is 10 minutes (10m).
```

Reply:
Missed the context timeout, thanks!